### DISCUSSION TITLE
notebook/stein_vgd

### DISCUSSION THREAD

**Discussion by giscus** [danmackinlay/danmackinlay.name-comments#2] [Comments]
# notebook/stein_vgd

KSD, SVGD, other computational Stein discrepancy methods

https://danmackinlay.name/notebook/stein_vgd.html

**Comment by aydinCkeser** [comment]
Excellent post Dan!
Here is my comment.

I take Eq.2 and multiply both sides with $p$:

$p A_p(f) = p f\cdot \nabla \log p + p \nabla \cdot f = f \cdot \nabla p + p\nabla \cdot f = \nabla \cdot (fp)$

That is a total derivative.

Now Eq.1 reads

$E[A_p(f)] = \int_V \nabla \cdot (fp) = \int_S p f\cdot \hat{n}$
by the divergence theorem.

Since we restricted $f$ such that $(f p) \to 0$ at infinity, the expectation value vanishes. However note that, for convergence, we need $f p$ to fall to zero at a sufficient rate.

If we compute the expectation numerically, we take a large volume in $X$, with radius $R$, therefore the surface integral in $n$ dimensions goes like

$R^{n-1} f(R b)p(R b)$ for $b$ a unit vector in $X$.

This means we not only require that $f p \to 0$ at $R\to\infty$ but also
$fp$ goes to zero faster than $1/R^{n-1}$, or more precisely:

 $R^{n-1} f(R|b|) p(R|b|) \to 0$   as $R \to \infty$ for  $b\in X, |b| =1$.

Maybe this is guaranteed by $p$ being a PDF with exponential tails otherwise we have an issue when we have many dimensions i.e. $n \gg 1$.

**Comment by aydinCkeser** [comment]
I have a second comment:

Above you choose $f$ to be a convex combination of basis functions $f_i$.

It is natural to take $f_i$ as tightly localized functions in $X$, for example $f_i = \mathcal{N}(\mu = X_i, \sigma^2)$.

It turns out this idea is not new at all.

So, inspired by physics I suggest waves as basis functions.

$f_{\mathbf{k}} \sim e^{-\alpha |\mathbf{x}|^2}\cos(\mathbf{k}\cdot \mathbf{x} + \theta_{\mathbf{k}})$
Here, $\alpha \to 0^+$ is a small positive number to ensure convergence.
Also, $k$ lives in the dual space (momentum space) of $X$ and can be enumerated in a discretization scheme.
The angles $\theta_\mathbf{k}$ can be freely chosen like the weights  $w_\mathbf{k}$, Indeed every weight comes like a pair $(w_\mathbf{k}, \theta_\mathbf{k})$.


If the probability is multimodal and  very delocalized, perhaps this might make the problem much easier. 

Actually one can generalize this concept. 
If the energy of your multiparticle ensemble is given,
I quantize it and use its eigenfunctions as basis for $f$.

What do you think?

**Reply by danmackinlay** [reply]
Interesting idea Aydin. I think we don't get to set $\theta_k$ completely freely without extra work, because then it leads to an uncountable basis, which means that we are no longer in a separable Hilbert space, which is an important technical condition. However, I think we can get arbitrarily close to that by letting the phase share the discretisation scheme of the $\mathbf{k}$, and working with functions like $\\{ e^{-\alpha|\mathbf{x}|^2} \sin \left(\mathbf{k} \cdot \mathbf{x}\right)\\}\_{\mathbf{k}}\cup \\{e^{-\alpha|\mathbf{x}|^2} \cos \left(\mathbf{k} \cdot \mathbf{x}\right)\\}\_{\mathbf{k}}$. The next question of course is what we do with them: are we talking about such functions as a “feature map”, with respect to which we define the kernel? or the kernel itself?

**Reply by aydinCkeser** [reply]
First part:

$\cos$ and $\sin$ are OK    , and I thought adding phases to $\cos$ is equivalent. If it is technically better to do it that way, so be it.

Second part:

I have not thought about the kernel part. 
My understanding is, we want to choose a natural set of basis functions, so that finding the coefficients $\omega_i$ that maximizes Stein discrepancy is a computationally feasible problem.

Is this correct?
